animations/hero-animations/index.md

```diff
@@ -0,0 +1,670 @@
+only the programmer needs to understand this implementation detail.
+
+<aside class="alert alert-info" markdown="1">
+**Note about dialogs:**
+Heroes fly from one PageRoute to another. Dialogs
+(displayed with `showDialog()`, for example), use PopupRoutes,
+which are not PageRoutes.  At least for now,
+you can't animate a hero to a Dialog.
+For further developments (and a possible workaround), [watch this
+issue.](https://github.com/flutter/flutter/issues/10667)
+</aside>
+
+Hero animation code has the following structure:
+
+1. Define a starting Hero object, referred to as the _source
+   hero_. The hero specifies its location,
+   size, graphical representation (typically an image), and an
+   identifying tag.
+1. Define an ending Hero object, referred to as the _destination hero_.
+   This hero also specifies its location, size, graphical
+   representation, and the same tag as the source hero.
+   It's <strong>essential that both hero objects are created with
+   the same tag</strong>, typically an object that represents the
+   underlying data. For best results, the heroes should have
+   virtually identical widget trees.
+1. Create a route that contains the destination hero.
+   The destination route defines the widget tree that exists
+   at the end of the animation.
+1. Trigger the animation by pushing the route on the Navigator's stack.
+   The Navigator push and pop operations trigger a hero animation
+   for each pair of heroes with matching tags in the source
+   and destination routes.
+
+Flutter calculates the tweens required to transition the
+starting hero to the ending hero (interpolating size and
+position changes), and performs the animation in an overlay.
+
+The next section describes Flutter's process in greater detail.
+
+## Behind the scenes
+
+The following describes how Flutter performs the transition.
+
+<img src="images/hero-transition-0.png" alt="Before the transition the source hero appears in the source route">
+
+Before transition, the source hero waits in the source route's widget
+tree. The destination route does not yet exist, and the overlay
+is empty.
+
+---
+
+<img src="images/hero-transition-1.png" alt="Transition begins and Destination hero appears in the overlay">
+
+Pushing a route to the Navigator triggers the animation.
+At t=0.0, Flutter does the following, mostly offscreen:
+
+* Calculates the transition (both position and size changes)
+  from beginning to end, moving the hero along a curved path as
+  described in the Material motion spec.
+  Flutter now knows where the hero ends up.
+
+* Places the destination hero in the overlay, at the
+  same location and size as the _source_ hero.
+  Adding a hero to the overlay changes its Z-order so that
+  it appears on top of all routes.
+
+* Moves the source hero offscreen.
+
+The framework kicks off the animation. When the animates
+completes (at t=1.0), Flutter moves both heroes back to the
```

Comment:
The material about what happens when the flight completes is repeated below. I don't think it's needed here.